reflex-dom-core/src/Reflex/Dom/Attributes/Types.hs

```diff
@@ -0,0 +1,298 @@
+
+lookupAttrInPatch :: AttributeName -> DeclareAttrs -> Maybe Text
+lookupAttrInPatch k (DeclareAttrs dm) = case DMap.lookup AttrKey_BasicAttributes dm of
+  Nothing -> Nothing
+  Just (PT (BasicAttributes m)) -> Map.lookup k m
+
+-- | Set an attribute to the given 'Text'.
+setAttribute :: AttributeName -> Text -> ModifyAttrs
+setAttribute k = ModifyAttrs . DMap.singleton AttrKey_BasicAttributes . Identity . ModBasicAttributes . PatchMap . Map.singleton k . Just
+
+-- | Remove an attribute.
+removeAttribute :: AttributeName -> ModifyAttrs
+removeAttribute k = ModifyAttrs . DMap.singleton AttrKey_BasicAttributes $ Identity $ ModBasicAttributes $ PatchMap $ Map.singleton k Nothing
+
+-- TODO we can't really remove anything except basic attributes, so some
+-- widgets which want total control over an attribute are possibly breakable
+removeAttrFromModifyAttrs :: AttributeName -> ModifyAttrs -> ModifyAttrs
+removeAttrFromModifyAttrs k = ModifyAttrs . DMap.alter f AttrKey_BasicAttributes . unModifyAttrs
+  where f = \case
+          Nothing -> Nothing
+          Just (Identity (ModBasicAttributes (PatchMap m))) -> Just $ Identity $ ModBasicAttributes $ PatchMap $ Map.delete k m
+
+-- TODO see comment for removeAttrFromModifyAttrs
+removeAttrFromDeclareAttrs :: AttributeName -> DeclareAttrs -> DeclareAttrs
+removeAttrFromDeclareAttrs k = DeclareAttrs . DMap.alter f AttrKey_BasicAttributes . unDeclareAttrs
+  where f = \case
+          Nothing -> Nothing
+          Just (PT (BasicAttributes m)) -> Just $ PT $ BasicAttributes $ Map.delete k m
+
+_PT :: Iso' (PT a) (PatchTarget a)
+_PT = iso (\(PT a) -> a) PT
+
+_BasicAttributes :: Iso' BasicAttributes (Map AttributeName Text)
+_BasicAttributes = iso unBasicAttributes BasicAttributes
+
+_DeclareAttrs :: Iso' DeclareAttrs (DMap AttrKey PT)
+_DeclareAttrs = iso unDeclareAttrs DeclareAttrs
+
+type instance Index DeclareAttrs = AttributeName
+type instance IxValue DeclareAttrs = Text
+instance Ixed DeclareAttrs where
+  ix k = _DeclareAttrs . DMap.dmix AttrKey_BasicAttributes . _PT . _BasicAttributes . ix k
+instance At DeclareAttrs where
+  at k f = _DeclareAttrs $ DMap.dmat AttrKey_BasicAttributes g
+    where g (Just (PT (BasicAttributes m))) = Just . PT . BasicAttributes <$> at k f m
+          g Nothing = Just . PT . BasicAttributes <$> at k f mempty
+
+_ModBasicAttributes :: Iso' ModBasicAttributes (Map AttributeName (Maybe Text))
+_ModBasicAttributes = iso (unPatchMap . unModBasicAttributes) (ModBasicAttributes . PatchMap)
+
+_ModifyAttrs :: Iso' ModifyAttrs (DMap AttrKey Identity)
+_ModifyAttrs = iso unModifyAttrs ModifyAttrs
+
+type instance Index ModifyAttrs = AttributeName
+type instance IxValue ModifyAttrs = Maybe Text
+instance Ixed ModifyAttrs where
+  ix k = _ModifyAttrs . DMap.dmix AttrKey_BasicAttributes . _Wrapped . _ModBasicAttributes . ix k
+instance At ModifyAttrs where
+  at k f = _ModifyAttrs $ DMap.dmat AttrKey_BasicAttributes g
```

Comment:
This instance seems like it would have the extremely confusing result that `(at "class" .~ Nothing) (addClass "hello")` would have an effect something like wiping out all the existing classes and then adding "hello", whereas `(at "class" .~ Nothing) $ "class" =: Just "hello"` would have the effect of removing all classes and not adding anything afterwards.